File: DuAnTauHoaCom/backend/src/controllers/TrainController.js
const TrainService = require('../services/TrainService')
const { ok, badRequest } = require('../utils/response')

const searchTrains = async (req, res, next) => {
  try {
    const { gaDi, gaDen, ngayChay } = req.query
    if (!gaDi || !gaDen || !ngayChay) return badRequest(res, 'Thiếu tham số: gaDi, gaDen, ngayChay')

    const results = await TrainService.searchTrains(gaDi, gaDen, ngayChay)
    ok(res, results)
  } catch (err) { next(err) }
}

const getSeatMap = async (req, res, next) => {
  try {
    const { idChuyen, soToa } = req.params
    const { idGaLen, idGaXuong } = req.query
    const result = await TrainService.getSeatMap(
      parseInt(idChuyen), parseInt(soToa),
      idGaLen  ? parseInt(idGaLen)  : null,
      idGaXuong ? parseInt(idGaXuong) : null
    )
    ok(res, result)
  } catch (err) { next(err) }
}

const getAllStations = async (req, res, next) => {
  try {
    const stations = await TrainService.getAllStations()
    ok(res, stations)
  } catch (err) { next(err) }
}

const getSchedule = async (req, res, next) => {
  try {
    const schedule = await TrainService.getSchedule()
    ok(res, schedule)
  } catch (err) { next(err) }
}

const getTrainDetail = async (req, res, next) => {
  try {
    const { idLichChay } = req.params
    const { idGaLen, idGaXuong, ngayChay } = req.query
    const result = await TrainService.getTrainRouteDetail(
      parseInt(idLichChay),
      idGaLen   ? parseInt(idGaLen)   : null,
      idGaXuong ? parseInt(idGaXuong) : null,
      ngayChay  || null
    )
    ok(res, result)
  } catch (err) { next(err) }
}

const getPopularRoutes = async (req, res, next) => {
  try {
    const limit = parseInt(req.query.limit) || 10
    const result = await TrainService.getPopularRoutes(limit)
    ok(res, result)
  } catch (err) { next(err) }
}

module.exports = { searchTrains, getSeatMap, getAllStations, getSchedule, getTrainDetail, getPopularRoutes }

File: java-backend/src/main/java/com/klntrain/controller/TrainController.java
package com.klntrain.controller;

import com.klntrain.dto.response.ApiResponse;
import com.klntrain.service.TrainService;
import com.klntrain.util.AppException;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/trains")
@RequiredArgsConstructor
public class TrainController {

    private final TrainService trainService;

    @GetMapping("/stations")
    public ResponseEntity<ApiResponse<?>> getAllStations() {
        return ResponseEntity.ok(ApiResponse.ok(trainService.getAllStations()));
    }

    @GetMapping("/search")
    public ResponseEntity<ApiResponse<?>> searchTrains(
        @RequestParam Long gaDi,
        @RequestParam Long gaDen,
        @RequestParam String ngayChay
    ) {
        if (gaDi == null || gaDen == null || ngayChay == null || ngayChay.isBlank()) {
            throw AppException.badRequest("Thiếu tham số: gaDi, gaDen, ngayChay");
        }
        var results = trainService.searchTrains(gaDi, gaDen, ngayChay);
        return ResponseEntity.ok(ApiResponse.ok(results));
    }

    @GetMapping("/{idChuyen}/seats/{soToa}")
    public ResponseEntity<ApiResponse<?>> getSeatMap(
        @PathVariable Long idChuyen,
        @PathVariable Integer soToa,
        @RequestParam(required = false) Long idGaLen,
        @RequestParam(required = false) Long idGaXuong
    ) {
        var result = trainService.getSeatMap(idChuyen, soToa, idGaLen, idGaXuong);
        return ResponseEntity.ok(ApiResponse.ok(result));
    }
}
